cranelift/codegen/src/isa/aarch64/lower.rs

```diff
@@ -1399,9 +1399,54 @@ pub(crate) fn lower_icmp<C: LowerCtx<I = Inst>>(
             }
         }
     } else if !ty.is_vector() {
-        let alu_op = choose_32_64(ty, ALUOp::SubS32, ALUOp::SubS64);
         let rn = put_input_in_reg(ctx, inputs[0], narrow_mode);
         let rm = put_input_in_rse_imm12(ctx, inputs[1], narrow_mode);
+
+        let is_overflow = condcode == IntCC::Overflow || condcode == IntCC::NotOverflow;
+        let is_small_type = ty == I8 || ty == I16;
+        let (rn, rm) = if is_overflow && is_small_type {
+            // Overflow checks for non native types require additional instructions, other than
+            // just the extend op.
+            //
+            // TODO: Codegen improvements here:
+            // * Merge the second sxt{h,b} into the sub instruction.
+            // * We can especially improve codegen here if we can return a different flag out of
+            // this function. That way we can tell the caller to use the 'ne' flag and save
+            // the last 3 instructions.
+            //
+            // sxt{h,b}  w0, w0
+            // sxt{h,b}  w1, w1
+            // sub       w0, w0, w1
+            // cmp       w0, w0, sxt{h,b}
+            // cset      w0, ne
+            // mov       w1, #0x80000000
+            // cmp       w1, w0
```

Comment:
Could you add some more explanation of this sequence?

I think I understand what's going on here -- we have a 0/1 flag generated by `cset` that is the actual overflow indicator, then we cause the main compare instruction to subtract that value from `INT32_MIN`, so we get an overflow (or underflow actually) on the subtract-as-compare. But it's somewhat difficult to work out!

Reply:
Yes,

The "main" overflow check is the `sub` and `cmp` with the extends, however this produces the result as a `ne` flag.

However the main issue is that the caller of this function is expecting to use the `vs` flag, and we (currently) don't have a way to tell it to use another flag. So we set the overflow flag by subtracting either 1 / 0 from INT32_MIN.

I'll update the comments to reflect this.

Reply:
FWIW, that sounds like a reasonable refactor, if you'd prefer to go that way: return the (possibly modified) condition flag and use that in the caller. I think that at every place where we use `lower_icmp`, we have full control over what condition code we actually use, so this should be pretty reasonable, and would improve the generated code.

Reply:
... and would also simplify the explanation that @cfallin has asked for :-). My vote would go towards refactoring as well.

Reply:
Ok, I'll update into that model instead.